rnd/autogpt_builder/src/components/CustomNode.tsx

```diff
@@ -281,10 +284,49 @@ export function CustomNode({ data, id }: NodeProps<CustomNode>) {
   }, [id, deleteElements]);
 
   const copyNode = useCallback(() => {
-    // This is a placeholder function. The actual copy functionality
-    // will be implemented by another team member.
-    console.log("Copy node:", id);
-  }, [id]);
+    const newId = Date.now().toString();
+    const currentNode = getNode(id);
+
+    if (!currentNode) {
+      console.error("Cannot copy node: current node not found");
+      return;
+    }
+
+    // Get the dimensions of the node
+    const nodeElement = document.querySelector(
+      `[data-id="${id}"]`,
+    ) as HTMLElement;
+    const nodeHeight = nodeElement ? nodeElement.offsetHeight : 0;
+    const nodeWidth = nodeElement ? nodeElement.offsetWidth : 0;
+
+    // Calculate new position with smart offset
+    const verticalOffset = nodeHeight + 20;
+
+    const newNode: CustomNode = {
+      id: newId,
+      type: "custom",
+      position: {
+        x: currentNode.position.x,
+        y: currentNode.position.y - verticalOffset,
+      },
+      data: {
+        ...data,
+        title: `${data.title} (Copy)`,
+        block_id: newId,
```

Comment:
`block_id` should carry over from the node being copied - it is the block type id e.g. `Printing Block`